File: wandb/sdk/internal/system/assets/trainium.py
import collections
import dataclasses
import json
import logging
import os
import pathlib
import shutil
import subprocess
import sys
import tempfile
import threading
import time
from collections import deque
from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Union

if sys.version_info >= (3, 8):
    from typing import Final
else:
    from typing_extensions import Final

from wandb.sdk.lib import telemetry

from .aggregators import aggregate_mean
from .asset_registry import asset_registry
from .interfaces import Interface, Metric, MetricsMonitor

if TYPE_CHECKING:
    from typing import Deque

    from wandb.sdk.internal.settings_static import SettingsStatic


logger = logging.getLogger(__name__)


NEURON_MONITOR_DEFAULT_CONFIG: Final[dict] = {
    "period": "1s",
    "neuron_runtimes": [
        {
            "tag_filter": ".*",
            "metrics": [
                {"type": "neuroncore_counters"},
                {"type": "memory_used"},
                {"type": "neuron_runtime_vcpu_usage"},
                # {"type": "execution_stats"},
            ],
        }
    ],
    "system_metrics": [
        {"type": "vcpu_usage"},
        {"type": "memory_info"},
        {"type": "neuron_hw_counters"},
    ],
}

# todo: once a python sdk is released with the Neuron utils, rewrite this
NEURON_LS_COMMAND: Final[Tuple[str, str]] = (
    shutil.which("neuron-ls") or "/opt/aws/neuron/bin/neuron-ls",
    "-j",
)
NEURON_MONITOR_PATH: Final[str] = (
    shutil.which("neuron-monitor") or "/opt/aws/neuron/bin/neuron-monitor"
)


@dataclasses.dataclass
class _NeuronCoreMemoryUsage:
    constants: int
    model_code: int
    model_shared_scratchpad: int
    runtime_memory: int
    tensors: int


@dataclasses.dataclass
class _HostMemoryUsage:
    application_memory: int
    constants: int
    dma_buffers: int
    tensors: int


@dataclasses.dataclass
class _Stats:
    neuroncore_utilization: Dict[int, float]  # per neuron core utilization
    host_total_memory_usage: int  # total memory usage in bytes
    neuron_device_total_memory_usage: int  # total memory usage
    host_memory_usage: _HostMemoryUsage  # host memory usage breakdown
    neuroncore_memory_usage: Dict[
        int, _NeuronCoreMemoryUsage
    ]  # per core memory usage breakdown


class NeuronCoreStats:
    """AWS Trainium stats."""

    name: str = "trn.{key}"
    samples: "Deque[_Stats]"

    def write_neuron_monitor_config(self) -> None:
        """Write neuron monitor config file."""
        # mkdir if not exists
        pathlib.Path(self.neuron_monitor_config_path).parent.mkdir(
            parents=True, exist_ok=True
        )
        # write default config
        with open(self.neuron_monitor_config_path, "w") as f:
            json.dump(NEURON_MONITOR_DEFAULT_CONFIG, f, indent=4)

    def neuron_monitor(self) -> None:
        """Run neuron-monitor in a separate process to collect raw data."""
        self.write_neuron_monitor_config()

        try:
            command = [
                NEURON_MONITOR_PATH,
                "-c",
                self.neuron_monitor_config_path,
            ]
            with subprocess.Popen(
                command,
                stdout=subprocess.PIPE,
                stderr=None,
            ) as process:
                while not self.shutdown_event.is_set():
                    if process.stdout is None:
                        self.shutdown_event.wait(0.1)
                        continue

                    raw_data = process.stdout.readline()
                    if raw_data:
                        self.raw_samples.append(raw_data)
                process.kill()
                process.wait()
        except Exception as e:
            logger.error("neuron-monitor failed: {}".format(e))

    def __init__(
        self,
        pid: int,
        neuron_monitor_config_path: Optional[str],
    ) -> None:
        self.pid = pid
        # neuron-monitor requires a config file (json)
        # we provide an option to supply a custom config file path
        # in case the default temp file path is not writable
        self.neuron_monitor_config_path = (
            neuron_monitor_config_path or tempfile.NamedTemporaryFile(delete=False).name
        )
        self.raw_samples: Deque[bytes] = deque(maxlen=10)
        self.samples: Deque[_Stats] = deque()
        self.shutdown_event = threading.Event()

        self.neuron_monitor_thread: Optional[threading.Thread] = None

    def setup(self) -> None:
        """Start the neuron-monitor thread for collecting raw data."""
        if self.neuron_monitor_thread is not None:
            return

        logger.debug("Starting neuron-monitor thread")
        self.shutdown_event.clear()
        self.neuron_monitor_thread = threading.Thread(
            name="NeuronCoreMntr",
            target=self.neuron_monitor,
            daemon=True,
        )
        self.neuron_monitor_thread.start()

    def teardown(self) -> None:
        """Stop the neuron-monitor thread."""
        logger.debug("Stopping neuron-monitor thread")
        try:
            self.shutdown_event.set()
            assert self.neuron_monitor_thread is not None
            self.neuron_monitor_thread.join()
        except Exception as e:
            logger.error("neuron-monitor thread failed to stop: {}".format(e))
        finally:
            self.neuron_monitor_thread = None

    def _is_matching_entry(self, entry: dict) -> bool:
        """Check if the entry should be saved.

        Checks if the pid in the entry matches the pid of the process.
        If not (as in the case of multi-process training with torchrun),
        checks if the LOCAL_RANK environment variable is set.

        todo: add matching by neuron_runtime_tag
        """
        return (int(entry["pid"]) == int(self.pid)) or "LOCAL_RANK" in os.environ

    def sample(self) -> None:
        try:
            raw_stats = json.loads(self.raw_samples[-1])
            neuron_runtime_data = [
                entry["report"]
                for entry in raw_stats["neuron_runtime_data"]
                if self._is_matching_entry(entry)
            ][0]  # there should be only one entry with the pid

            neuroncores_in_use = neuron_runtime_data["neuroncore_counters"][
                "neuroncores_in_use"
            ]
            # per-core utilization stats:
            neuroncore_utilization = {
                int(k): v["neuroncore_utilization"]
                for k, v in neuroncores_in_use.items()
            }
            # memory usage
            neuron_runtime_used_bytes = neuron_runtime_data["memory_used"][
                "neuron_runtime_used_bytes"
            ]
            # memory usage totals
            host_total_memory_usage = neuron_runtime_used_bytes["host"]
            neuron_device_total_memory_usage = neuron_runtime_used_bytes[
                "neuron_device"
            ]
            # memory usage breakdown
            usage_breakdown = neuron_runtime_used_bytes["usage_breakdown"]
            host_memory_usage = _HostMemoryUsage(**usage_breakdown["host"])
            neuroncore_memory_usage = {
                int(k): _NeuronCoreMemoryUsage(**v)
                for k, v in usage_breakdown["neuroncore_memory_usage"].items()
            }

            # When the training script is executed with torchrun,
            # we only want to keep the relevant LOCAL_RANK stats
            local_rank = int(os.environ.get("LOCAL_RANK", -1337))
            if local_rank >= 0:
                neuroncore_utilization = {
                    local_rank: neuroncore_utilization[local_rank]
                }
                neuroncore_memory_usage = {
                    local_rank: neuroncore_memory_usage[local_rank]
                }

            stats: _Stats = _Stats(
                neuroncore_utilization=neuroncore_utilization,
                host_total_memory_usage=host_total_memory_usage,
                neuron_device_total_memory_usage=neuron_device_total_memory_usage,
                host_memory_usage=host_memory_usage,
                neuroncore_memory_usage=neuroncore_memory_usage,
            )
            self.samples.append(stats)

        except Exception as e:  # noqa
            pass

    def clear(self) -> None:
        self.samples.clear()

    @staticmethod
    def flatten_stats(sample: _Stats) -> dict:
        """Flatten _Stats object into a flat dict of numbers."""
        flattened = {}

        def helper(key: str, value: Any) -> None:
            if isinstance(value, (int, float)):
                ret = {f"{key}": value}
                flattened.update(ret)
                return
            elif isinstance(value, dict):
                for kk, vv in value.items():
                    if isinstance(kk, int):
                        # top-level keys are neuron core ids,
                        # so we swap the order to comply with the
                        # frontend expectations
                        helper(f"{kk}.{key}", vv)
                    else:
                        helper(f"{key}.{kk}", vv)
                return
            elif isinstance(value, list):
                for i, val in enumerate(value):
                    helper(f"{i}.{key}", val)

        for kkk, vvv in dataclasses.asdict(sample).items():
            helper(kkk, vvv)

        return flattened

    def aggregate(self) -> dict:
        if not self.samples:
            return {}

        stats = {}

        # Stats could be: numbers or dataclass objects or lists of such.
        # In the latter case that means per-core stats.
        # The dataclass objects are flat containers of numbers.

        # flatten samples and merge the corresponding values into lists
        merged_samples: Dict[str, List[Union[int, float]]] = collections.defaultdict(
            list
        )
        for flattened_sample in (self.flatten_stats(sample) for sample in self.samples):
            for k, v in flattened_sample.items():
                merged_samples[k].append(v)

        # aggregate the lists
        for k, v in merged_samples.items():
            stats[self.name.format(key=k)] = aggregate_mean(v)

        return stats


@asset_registry.register
class Trainium:
    def __init__(
        self,
        interface: "Interface",
        settings: "SettingsStatic",
        shutdown_event: threading.Event,
    ) -> None:
        self.name = self.__class__.__name__.lower()
        self.metrics: List[Metric] = [
            NeuronCoreStats(
                settings._stats_pid,
                settings._stats_neuron_monitor_config_path,
            ),
        ]
        self.metrics_monitor = MetricsMonitor(
            self.name,
            self.metrics,
            interface,
            settings,
            shutdown_event,
        )
        telemetry_record = telemetry.TelemetryRecord()
        telemetry_record.env.trainium = True
        interface._publish_telemetry(telemetry_record)

    @classmethod
    def is_available(cls) -> bool:
        # todo: check if neuron-ls is available and if yes, what it reports. see:
        # https://awsdocs-neuron.readthedocs-hosted.com/en/latest/tools/neuron-sys-tools/neuron-ls.html
        if not pathlib.Path(NEURON_LS_COMMAND[0]).exists():
            return False
        # need to be extra careful as neuron tools could be pre-installed
        # on some systems that do not have the hardware
        try:
            # redirect stderr to null to avoid printing errors to the console
            # todo: alternative: check /dev/neuron0 ? sysfs support coming soon in neuron tools
            output = subprocess.check_output(
                NEURON_LS_COMMAND,
                universal_newlines=True,
                stderr=subprocess.DEVNULL,
            ).strip()
            if len(json.loads(output)) > 0:
                return True
        except (OSError, ValueError, TypeError, subprocess.CalledProcessError):
            pass

        return False

    def start(self) -> None:
        self.metrics_monitor.start()

    def finish(self) -> None:
        self.metrics_monitor.finish()

    def probe(self) -> dict:
        try:
            self.metrics[0].check_neuron_monitor_config()  # type: ignore
            neuron_hardware_info: dict = {}
            command = [
                NEURON_MONITOR_PATH,
                "-c",
                self.metrics[0].neuron_monitor_config_path,  # type: ignore
            ]
            with subprocess.Popen(
                command,
                stdout=subprocess.PIPE,
                stderr=None,
            ) as process:
                while True:
                    if process.stdout is None:
                        time.sleep(0.1)
                        continue

                    raw_data = process.stdout.readline()
                    if raw_data:
                        parsed_data = json.loads(raw_data)
                        neuron_hardware_info = parsed_data.get(
                            "neuron_hardware_info", {}
                        )
                        neuron_hardware_info.pop("error", None)
                        break

            try:
                process.kill()
                process.wait()
            except:  # noqa
                pass

            return {self.name: neuron_hardware_info}
        except Exception as e:
            logger.error("neuron-monitor failed: {}".format(e))
            return {}

File: core/pkg/monitor/trainium.go
//go:build linux

package monitor

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"sync"
	"time"

	"github.com/wandb/wandb/core/internal/observability"
	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
)

// NeuronMonitorConfig represents the configuration for the neuron-monitor command.
type NeuronMonitorConfig struct {
	Period         string                `json:"period"`
	NeuronRuntimes []NeuronRuntimeConfig `json:"neuron_runtimes"`
	SystemMetrics  []SystemMetricConfig  `json:"system_metrics"`
}

type NeuronRuntimeConfig struct {
	TagFilter string         `json:"tag_filter"`
	Metrics   []MetricConfig `json:"metrics"`
}

type SystemMetricConfig struct {
	Type string `json:"type"`
}

type MetricConfig struct {
	Type string `json:"type"`
}

// NeuronCoreMemoryUsage represents the memory usage breakdown for a neuron core.
type NeuronCoreMemoryUsage struct {
	Constants             int `json:"constants"`
	ModelCode             int `json:"model_code"`
	ModelSharedScratchpad int `json:"model_shared_scratchpad"`
	RuntimeMemory         int `json:"runtime_memory"`
	Tensors               int `json:"tensors"`
}

// HostMemoryUsage represents the memory usage breakdown on the host.
type HostMemoryUsage struct {
	ApplicationMemory int `json:"application_memory"`
	Constants         int `json:"constants"`
	DmaBuffers        int `json:"dma_buffers"`
	Tensors           int `json:"tensors"`
}

// TrainiumStats represents the stats returned by the neuron-monitor command.
//
// NeuroncoreUtilization: per neuron core utilization
// HostTotalMemoryUsage: total memory usage in bytes
// NeuronDeviceTotalMemoryUsage: total memory usage on neuron device in bytes
// HostMemoryUsage: host memory usage breakdown
// NeuroncoreMemoryUsage: per neuron core memory usage breakdown
type TrainiumStats struct {
	NeuroncoreUtilization        map[int]float64               `json:"neuroncore_utilization"`
	HostTotalMemoryUsage         int                           `json:"host_total_memory_usage"`
	NeuronDeviceTotalMemoryUsage int                           `json:"neuron_device_total_memory_usage"`
	HostMemoryUsage              HostMemoryUsage               `json:"host_memory_usage"`
	NeuroncoreMemoryUsage        map[int]NeuronCoreMemoryUsage `json:"neuroncore_memory_usage"`
}

// Trainium is a monitor for AWS Trainium devices.
//
// Uses the neuron-monitor command to get stats.
type Trainium struct {
	name                    string
	pid                     int32
	samplingInterval        float64
	neuronMonitorConfigPath string
	mutex                   sync.RWMutex
	cmd                     *exec.Cmd
	logger                  *observability.CoreLogger
	rawStats                map[string]any
	shutdownEvent           chan struct{}
	isRunning               bool
}

// getCmdPath returns the path to the neuron-monitor command.
func getNeuronMonitorCmdPath() (string, error) {
	// try to find the command in the PATH
	exPath, err := exec.LookPath("neuron-monitor")
	if err == nil {
		return exPath, nil
	}
	// try the default path
	exPath = "/opt/aws/neuron/bin/neuron-monitor"
	if _, err := os.Stat(exPath); os.IsNotExist(err) {
		return "", err
	}
	return exPath, nil
}

func NewTrainium(
	logger *observability.CoreLogger,
	pid int32,
	samplingInterval float64,
	neuronMonitorConfigPath string,
) *Trainium {
	t := &Trainium{
		name:                    "trainium",
		pid:                     pid,
		samplingInterval:        samplingInterval,
		neuronMonitorConfigPath: neuronMonitorConfigPath,
		logger:                  logger,
		shutdownEvent:           make(chan struct{}),
	}

	// check if the neuron-monitor command is available
	if _, err := getNeuronMonitorCmdPath(); err != nil {
		return nil
	}

	if t.samplingInterval == 0 {
		t.samplingInterval = 1.0
	}

	// neuron-monitor requires a JSON config file.
	// we provide an option to supply a custom config file path
	// in case the default temp file path is not writable.
	if t.neuronMonitorConfigPath == "" {
		t.neuronMonitorConfigPath = filepath.Join(os.TempDir(), "neuron_monitor_config.json")
		err := t.writeNeuronMonitorConfig(t.neuronMonitorConfigPath)
		if err != nil {
			return nil
		}
	}

	err := t.Start()
	if err != nil {
		return nil
	}

	return t
}

// writeNeuronMonitorConfig writes the neuron-monitor config to a file.
func (t *Trainium) writeNeuronMonitorConfig(neuronMonitorConfigPath string) error {
	config := NeuronMonitorConfig{
		Period: fmt.Sprintf("%ds", int(t.samplingInterval)),
		NeuronRuntimes: []NeuronRuntimeConfig{
			{
				TagFilter: ".*",
				Metrics: []MetricConfig{
					{Type: "neuroncore_counters"},
					{Type: "memory_used"},
					{Type: "neuron_runtime_vcpu_usage"},
				},
			},
		},
		SystemMetrics: []SystemMetricConfig{
			{Type: "vcpu_usage"},
			{Type: "memory_info"},
			{Type: "neuron_hw_counters"},
		},
	}

	jsonData, err := json.MarshalIndent(config, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal config: %v", err)
	}

	err = os.WriteFile(neuronMonitorConfigPath, jsonData, 0644)
	if err != nil {
		return fmt.Errorf("failed to write config file: %v", err)
	}

	return nil
}

func (t *Trainium) SetRawStats(rawStats map[string]any) {
	t.mutex.Lock()
	t.rawStats = rawStats
	t.mutex.Unlock()
}

func (t *Trainium) SetRunningState(running bool) {
	t.isRunning = running
}

// Start executes the neuron-monitor command and reads its output in a separate goroutine.
//
// The output is expected to be JSON. It is parsed and stored in the rawStats field.
func (t *Trainium) Start() error {
	if t.isRunning {
		return fmt.Errorf("Trainium monitor is already running")
	}

	exPath, err := getNeuronMonitorCmdPath()
	if err != nil {
		return fmt.Errorf("failed to get command path: %v", err)
	}

	t.cmd = exec.Command(exPath, "-c", t.neuronMonitorConfigPath)

	stdout, err := t.cmd.StdoutPipe()
	if err != nil {
		return fmt.Errorf("failed to get stdout pipe: %v", err)
	}

	if err := t.cmd.Start(); err != nil {
		return fmt.Errorf("failed to start command: %v", err)
	}

	t.SetRunningState(true)

	go func() {
		scanner := bufio.NewScanner(stdout)
		for {
			select {
			case <-t.shutdownEvent:
				return
			default:
				if scanner.Scan() {
					rawStats := make(map[string]any)
					if err := json.Unmarshal(scanner.Bytes(), &rawStats); err != nil {
						t.logger.CaptureError(fmt.Errorf("trainium: failed to parse JSON: %v", err))
						continue
					}
					t.SetRawStats(rawStats)
				}
			}
		}
	}()

	return nil
}

// isMatchingEntry checks if an entry in neuronRuntimeData should be saved.
//
// Checks if the pid in the entry matches the pid of the process.
// If not (as in the case of multi-process training with torchrun),
// checks if the LOCAL_RANK environment variable is set.
//
// TODO: add matching by neuron_runtime_tag
func (t *Trainium) isMatchingEntry(entry map[string]any) bool {
	entryPid, ok := entry["pid"].(float64)
	if !ok {
		return false
	}
	return int32(entryPid) == t.pid || os.Getenv("LOCAL_RANK") != ""
}

// Sample returns the latest stats from the neuron-monitor command.
//
// The stats are parsed into a TrainiumStats struct, flattened and returned as a map.
//
//gocyclo:ignore
func (t *Trainium) Sample() (map[string]any, error) {
	if !t.isRunning {
		return nil, nil
	}

	t.mutex.RLock()
	rawStats := t.rawStats
	t.mutex.RUnlock()

	neuronRuntimeData, ok := rawStats["neuron_runtime_data"].([]any)
	if !ok || len(neuronRuntimeData) == 0 {
		return nil, nil
	}

	var matchingEntry map[string]any
	for _, entry := range neuronRuntimeData {
		if entryMap, ok := entry.(map[string]any); ok {
			if t.isMatchingEntry(entryMap) {
				matchingEntry = entryMap
				break
			}
		}
	}

	if matchingEntry == nil {
		return nil, nil
	}

	report, ok := matchingEntry["report"].(map[string]any)
	if !ok {
		return nil, nil
	}

	neuroncoreCounters, ok := report["neuroncore_counters"].(map[string]any)
	if !ok {
		return nil, nil
	}

	neuronCoresInUse, ok := neuroncoreCounters["neuroncores_in_use"].(map[string]any)
	if !ok {
		return nil, nil
	}

	neuroncoreUtilization := make(map[int]float64)
	for k, v := range neuronCoresInUse {
		coreID, _ := strconv.Atoi(k)
		if coreData, ok := v.(map[string]any); ok {
			if utilization, ok := coreData["neuroncore_utilization"].(float64); ok {
				neuroncoreUtilization[coreID] = utilization
			}
		}
	}

	memoryUsed, ok := report["memory_used"].(map[string]any)
	if !ok {
		return nil, nil
	}

	neuronRuntimeUsedBytes, ok := memoryUsed["neuron_runtime_used_bytes"].(map[string]any)
	if !ok {
		return nil, nil
	}

	hostTotalMemoryUsage, _ := neuronRuntimeUsedBytes["host"].(float64)
	neuronDeviceTotalMemoryUsage, _ := neuronRuntimeUsedBytes["neuron_device"].(float64)

	usageBreakdown, ok := neuronRuntimeUsedBytes["usage_breakdown"].(map[string]any)
	if !ok {
		return nil, nil
	}

	var hostMemoryUsage HostMemoryUsage
	if hostUsage, ok := usageBreakdown["host"].(map[string]any); ok {
		jsonBytes, err := json.Marshal(hostUsage)
		if err == nil {
			err = json.Unmarshal(jsonBytes, &hostMemoryUsage)
		}
		if err != nil {
			t.logger.CaptureError(fmt.Errorf("trainium: failed to unmarshal host memory usage: %v", err))
		}
	}

	neuroncoreMemoryUsage := make(map[int]NeuronCoreMemoryUsage)
	if ncMemUsage, ok := usageBreakdown["neuroncore_memory_usage"].(map[string]any); ok {
		for k, v := range ncMemUsage {
			coreID, _ := strconv.Atoi(k)
			jsonBytes, err := json.Marshal(v)
			if err == nil {
				var coreUsage NeuronCoreMemoryUsage
				err = json.Unmarshal(jsonBytes, &coreUsage)
				if err == nil {
					neuroncoreMemoryUsage[coreID] = coreUsage
				}
			}
			if err != nil {
				t.logger.CaptureError(fmt.Errorf("trainium: failed to unmarshal neuroncore memory usage: %v", err))
			}
		}
	}

	// When the training script is executed with torchrun,
	// we only want to keep the relevant LOCAL_RANK stats
	localRank := os.Getenv("LOCAL_RANK")
	if localRank != "" {
		localRankInt, _ := strconv.Atoi(localRank)
		neuroncoreUtilization = map[int]float64{localRankInt: neuroncoreUtilization[localRankInt]}
		neuroncoreMemoryUsage = map[int]NeuronCoreMemoryUsage{localRankInt: neuroncoreMemoryUsage[localRankInt]}
	}

	stats := TrainiumStats{
		NeuroncoreUtilization:        neuroncoreUtilization,
		HostTotalMemoryUsage:         int(hostTotalMemoryUsage),
		NeuronDeviceTotalMemoryUsage: int(neuronDeviceTotalMemoryUsage),
		HostMemoryUsage:              hostMemoryUsage,
		NeuroncoreMemoryUsage:        neuroncoreMemoryUsage,
	}

	return t.flattenStats(stats), nil
}

// flattenStats recursively flattens the stats into a map.
//
// Keys are prepended with "trn." to be recognized by the frontend.
func (t *Trainium) flattenStats(sample TrainiumStats) map[string]any {
	flattened := make(map[string]any)

	var flatten func(string, any)
	flatten = func(key string, value any) {
		switch v := value.(type) {
		case int:
			flattened[key] = float64(v)
		case float64:
			flattened[key] = v
		case map[int]float64:
			for k, vv := range v {
				flatten(fmt.Sprintf("%d.%s", k, key), vv)
			}
		case map[int]NeuronCoreMemoryUsage:
			for k, vv := range v {
				flatten(fmt.Sprintf("%d.%s", k, key), vv)
			}
		case HostMemoryUsage, NeuronCoreMemoryUsage:
			jsonBytes, _ := json.Marshal(v)
			var subMap map[string]any
			err := json.Unmarshal(jsonBytes, &subMap)
			if err != nil {
				t.logger.CaptureError(fmt.Errorf("trainium: failed to unmarshal submap: %v", err))
				return
			}
			for subKey, subValue := range subMap {
				flatten(fmt.Sprintf("%s.%s", key, subKey), subValue)
			}
		}
	}

	flatten("neuroncore_utilization", sample.NeuroncoreUtilization)
	flatten("host_total_memory_usage", sample.HostTotalMemoryUsage)
	flatten("neuron_device_total_memory_usage", sample.NeuronDeviceTotalMemoryUsage)
	flatten("host_memory_usage", sample.HostMemoryUsage)
	flatten("neuroncore_memory_usage", sample.NeuroncoreMemoryUsage)

	// Prepend "trn." to each key. This is necessary for the frontend to recognize the keys.
	result := make(map[string]any, len(flattened))
	for k, v := range flattened {
		newKey := "trn." + k
		result[newKey] = v
	}

	return result
}

func (t *Trainium) Name() string {
	return t.name
}

func (t *Trainium) IsAvailable() bool {
	_, err := getNeuronMonitorCmdPath()
	return err == nil
}

// Close stops the neuron-monitor command and sets isRunning to false.
func (t *Trainium) Close() {
	if !t.isRunning {
		return
	}

	close(t.shutdownEvent)
	if t.cmd != nil && t.cmd.Process != nil {
		err := t.cmd.Process.Kill()
		if err != nil {
			t.logger.CaptureError(fmt.Errorf("trainium: failed to kill process: %v", err))
		}
	}
	t.SetRunningState(false)
}

func (t *Trainium) Probe() *spb.MetadataRequest {
	if !t.IsAvailable() {
		return nil
	}

	info := &spb.MetadataRequest{
		Trainium: &spb.TrainiumInfo{
			Name:   t.name,
			Vendor: "AWS",
		},
	}

	// Wait for the first sample, but no more than 5 seconds.
	startTime := time.Now()
	for {
		t.mutex.RLock()
		_, ok := t.rawStats["neuron_hardware_info"]
		t.mutex.RUnlock()
		if ok {
			break // Successfully got a sample
		}
		if time.Since(startTime) > 5*time.Second {
			// just give up if we don't get a sample in 5 seconds
			return nil
		}
		time.Sleep(100 * time.Millisecond)
	}

	neuronHardwareInfo, ok := t.rawStats["neuron_hardware_info"].(map[string]any)
	if !ok {
		return nil
	}

	neuronDeviceCount, ok := neuronHardwareInfo["neuron_device_count"].(uint32)
	if ok {
		info.Trainium.NeuronDeviceCount = neuronDeviceCount
	}
	neuroncorePerDeviceCount, ok := neuronHardwareInfo["neuroncore_per_device_count"].(uint32)
	if ok {
		info.Trainium.NeuroncorePerDeviceCount = neuroncorePerDeviceCount
	}

	return info
}
